File: frontend/lib/dashboard-db.ts
import mysql from 'mysql2/promise';

const dbName = process.env.DB_NAME || 'factory';

function getConfig() {
  return {
    host: process.env.DB_HOST || 'localhost',
    port: Number(process.env.DB_PORT || 3306),
    user: process.env.DB_USER || 'root',
    password: process.env.DB_PASSWORD || '',
    database: dbName,
    connectTimeout: 10000,
  };
}

export async function getConnection() {
  return mysql.createConnection(getConfig());
}

/** MariaDB/MySQL 식별자(테이블·컬럼명) 이스케이프. 공백·특수문자·백틱 포함 시에도 안전 */
export function escapeSqlId(name: string): string {
  const safe = String(name).replace(/`/g, '``');
  return '`' + safe + '`';
}

/** 동적 SQL에 넣을 컬럼명이 안전한지. 영문·숫자·밑줄·공백만 허용 (^, 따옴표, 백틱 등 제외) */
export function isSafeColumnName(name: string): boolean {
  return typeof name === 'string' && name.length > 0 && /^[a-zA-Z0-9_\s]+$/.test(name);
}

/** 테이블 목록 조회 */
export async function getTables(conn: mysql.Connection): Promise<string[]> {
  const [rows] = await conn.query<any[]>('SHOW TABLES');
  const key = rows?.length ? Object.keys(rows[0])[0] : null;
  return key ? rows.map((r) => r[key]).filter(Boolean) : [];
}

/** 테이블 컬럼 정보 (이름, 타입) */
export async function getColumns(
  conn: mysql.Connection,
  table: string
): Promise<{ name: string; type: string }[]> {
  const [rows] = await conn.query<any[]>(
    `SELECT COLUMN_NAME as name, DATA_TYPE as type
     FROM information_schema.COLUMNS
     WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
     ORDER BY ORDINAL_POSITION`,
    [dbName, table]
  );
  return rows || [];
}

/** 날짜형 컬럼명 후보 (simulation_results 기준: timestamp 우선) */
const DATE_COLUMNS = ['timestamp', 'date', 'created_at', 'recorded_at', 'dt', 'record_date'];
/** 수치형 컬럼명 후보 (생산량 등) */
const VALUE_COLUMNS = ['quantity', 'amount', 'count', 'qty', 'value', 'production', 'total'];
/** 합계용 수치 컬럼 (에너지 등) */
const SUM_COLUMNS = ['consumption', 'usage', 'kwh', 'value', 'amount', 'cost'];

function pickColumn(columns: { name: string; type: string }[], candidates: string[]) {
  const names = columns.map((c) => c.name.toLowerCase());
  for (const c of candidates) {
    if (names.includes(c.toLowerCase())) return c;
    const found = columns.find((col) => col.name.toLowerCase().includes(c));
    if (found) return found.name;
  }
  return null;
}

function isDateType(type: string) {
  const t = (type || '').toLowerCase();
  return t.includes('date') || t.includes('time') || t === 'timestamp';
}

function isNumericType(type: string) {
  const t = (type || '').toLowerCase();
  return t.includes('int') || t.includes('decimal') || t.includes('float') || t.includes('double');
}

/** 테이블에서 날짜 컬럼 1개, 수치 컬럼 1개 찾기 */
export function findDateAndValueColumns(columns: { name: string; type: string }[]) {
  const dateCol =
    pickColumn(columns, DATE_COLUMNS) ||
    columns.find((c) => isDateType(c.type))?.name ||
    null;
  const valueCol =
    pickColumn(columns, VALUE_COLUMNS) ||
    pickColumn(columns, SUM_COLUMNS) ||
    columns.find((c) => isNumericType(c.type))?.name ||
    null;
  return { dateCol, valueCol };
}

/** 기간 조건 (day: 오늘, week: 최근 7일, month: 최근 4주) */
export function periodCondition(
  dateCol: string,
  period: 'day' | 'week' | 'month'
): string {
  const col = dateCol || 'timestamp';
  if (period === 'day') return `DATE(\`${col}\`) = ?`;
  if (period === 'week') return `\`${col}\` >= DATE_SUB(CURDATE(), INTERVAL 7 DAY)`;
  return `\`${col}\` >= DATE_SUB(CURDATE(), INTERVAL 28 DAY)`;
}

/** 오늘 날짜 (YYYY-MM-DD). Vercel 서버 UTC 대비 DB가 KST면 BACKEND_DATE_TZ=Asia/Seoul 사용 */
export function getTodayDateString(): string {
  const tz = process.env.BACKEND_DATE_TZ || 'Asia/Seoul';
  try {
    const s = new Date().toLocaleDateString('en-CA', { timeZone: tz });
    if (/^\d{4}-\d{2}-\d{2}$/.test(s)) return s;
  } catch {
    // fallback: 서버 로컬
  }
  const n = new Date();
  const y = n.getFullYear();
  const m = String(n.getMonth() + 1).padStart(2, '0');
  const d = String(n.getDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}

/** LOT별 공정현황용 날짜 문자열 (todayStr 기준, 주/월 동일 달력) */
export function getDashboardDateStrings(): {
  todayStr: string;
  weekStartStr: string;
  weekEndStr: string;
  firstOfMonth: string;
  lastOfMonthStr: string;
} {
  const todayStr = getTodayDateString();
  const [y, m, d] = todayStr.split('-').map(Number);
  const pad = (n: number) => String(n).padStart(2, '0');
  const dayOfWeek = new Date(Date.UTC(y, m - 1, d)).getUTCDay();
  const monOffset = (dayOfWeek + 6) % 7;
  const monDate = new Date(Date.UTC(y, m - 1, d - monOffset));
  const sunDate = new Date(monDate);
  sunDate.setUTCDate(monDate.getUTCDate() + 6);
  const weekStartStr = `${monDate.getUTCFullYear()}-${pad(monDate.getUTCMonth() + 1)}-${pad(monDate.getUTCDate())}`;
  const weekEndStr = `${sunDate.getUTCFullYear()}-${pad(sunDate.getUTCMonth() + 1)}-${pad(sunDate.getUTCDate())}`;
  const firstOfMonth = `${y}-${pad(m)}-01`;
  const lastDay = new Date(Date.UTC(y, m, 0)).getUTCDate();
  const lastOfMonthStr = `${y}-${pad(m)}-${pad(lastDay)}`;
  return { todayStr, weekStartStr, weekEndStr, firstOfMonth, lastOfMonthStr };
}

/** 공정 데이터용 메인 테이블. simulation_results 우선 (prediction 컬럼 있음) */
export async function getProcessDataTable(conn: mysql.Connection): Promise<string> {
  const fixed = process.env.PROCESS_TABLE_NAME?.trim();
  if (fixed) return fixed;
  const tables = await getTables(conn);
  const lower = (t: string) => t.toLowerCase();
  // simulation_results 테이블 우선 (prediction 컬럼 있음)
  const found =
    tables.find((t) => lower(t) === 'simulation_results') ??
    tables.find((t) => lower(t) === 'preprocessing') ??
    tables.find((t) => lower(t) === 'raw_data') ??
    tables.find((t) => lower(t) === 'defect_results') ??
    tables[0];
  return found ?? 'simulation_results';
}

/** 컬럼명이 후보 중 하나를 포함하거나 일치하면 해당 컬럼명 반환 */
function findColumn(columns: { name: string; type: string }[], candidates: string[]): string | null {
  const lower = (s: string) => s.toLowerCase();
  for (const cand of candidates) {
    const exact = columns.find((c) => lower(c.name) === lower(cand));
    if (exact) return exact.name;
    const partial = columns.find((c) => lower(c.name).includes(lower(cand)) || lower(cand).includes(lower(c.name)));
    if (partial) return partial.name;
  }
  return null;
}

/** 날짜형 컬럼 찾기. simulation_results 기준: timestamp 우선 */
function findDateColumn(columns: { name: string; type: string }[]): string | null {
  const dateNames = ['timestamp', 'date', 'created_at', 'recorded_at', 'dt', 'time', '날짜'];
  const byName = findColumn(columns, dateNames);
  if (byName) return byName;
  return columns.find((c) => isDateType(c.type))?.name || null;
}

/** 공정 테이블 한 개의 컬럼 매핑: 생산/품질/에너지/설비 등에 쓸 컬럼 자동 감지 */
export interface ProcessColumnMap {
  table: string;
  dateCol: string | null;
  quantityCol: string | null;
  passRateCol: string | null;
  defectCol: string | null;
  consumptionCol: string | null;
  efficiencyCol: string | null;
  lineCol: string | null;
  /** LOT 식별용 컬럼 (lot_id, lot, batch, id 등) */
  lotCol: string | null;
  /** 합불여부 컬럼 (pass/fail, 합격/불량 등) */
  resultCol: string | null;
  /** 기타 수치 컬럼들 (대시보드 추가 KPI용) */
  numericCols: string[];
}

export async function getProcessColumnMap(
  conn: mysql.Connection,
  tableName: string
): Promise<ProcessColumnMap> {
  const columns = await getColumns(conn, tableName);
  const dateCol = findDateColumn(columns);
  const quantityCol = findColumn(columns, ['quantity', 'amount', 'count', 'qty', 'output', '생산', '수량']);
  const passRateCol = findColumn(columns, ['pass_rate', 'pass', 'quality', 'ok_rate', '양품률', '품질']);
  const defectCol = findColumn(columns, ['prediction', 'quality_defect', 'defect', 'defect_rate', 'fail', 'ng', '불량']);
  const consumptionCol = findColumn(columns, ['consumption', 'usage', 'kwh', 'energy', 'power', '에너지', '소비']);
  const efficiencyCol = findColumn(columns, ['efficiency', 'uptime', 'oee', 'rate', '효율', '가동률']);
  const lineCol = findColumn(columns, ['line', 'line_id', 'line_name', '라인', '공정']);
  const lotCol = findColumn(columns, ['lot_id', 'lot', 'batch', 'lot_no', 'batch_id', 'LOT', 'id']);
  const resultCol = findColumn(columns, ['prediction', 'quality_defect', 'y_defect', 'result', 'pass_fail', 'judge', 'judgment', '판정', '합불', 'ok_ng', 'pass_fail_result', 'quality_result', 'judgement']);

  const numericCols = columns
    .filter((c) => isNumericType(c.type))
    .map((c) => c.name);

  const quantityFallback = quantityCol || numericCols.find((n) => !/pass|rate|quality|efficiency|uptime|defect/i.test(n)) || numericCols[0] || null;

  return {
    table: tableName,
    dateCol,
    quantityCol: quantityFallback,
    passRateCol,
    defectCol,
    consumptionCol,
    efficiencyCol,
    lineCol,
    lotCol,
    resultCol,
    numericCols,
  };
}

File: backend/backend_fastapi/dashboard_db.py
"""공정 대시보드용 DB 헬퍼 (Next.js lib/dashboard-db.ts 포팅)."""
import re
from datetime import datetime
from zoneinfo import ZoneInfo

from config import PROCESS_DB, BACKEND_DATE_TZ
from db import get_process_connection


def escape_sql_id(name: str) -> str:
    safe = str(name).replace("`", "``")
    return f"`{safe}`"


def is_safe_column_name(name: str) -> bool:
    return bool(name and re.match(r"^[a-zA-Z0-9_\s]+$", name))


def get_tables(conn) -> list:
    with conn.cursor() as cur:
        cur.execute("SHOW TABLES")
        rows = cur.fetchall()
    if not rows:
        return []
    key = list(rows[0].keys())[0]
    return [r[key] for r in rows if r.get(key)]


def get_columns(conn, table: str) -> list[dict]:
    db = PROCESS_DB["database"]
    with conn.cursor() as cur:
        cur.execute(
            """SELECT COLUMN_NAME as name, DATA_TYPE as type
               FROM information_schema.COLUMNS
               WHERE TABLE_SCHEMA = %s AND TABLE_NAME = %s
               ORDER BY ORDINAL_POSITION""",
            (db, table),
        )
        return cur.fetchall() or []


def _pick_column(columns: list[dict], candidates: list[str]) -> str | None:
    names = [c["name"].lower() for c in columns]
    for cand in candidates:
        if cand.lower() in names:
            for c in columns:
                if c["name"].lower() == cand.lower():
                    return c["name"]
        for c in columns:
            if cand.lower() in c["name"].lower() or c["name"].lower() in cand.lower():
                return c["name"]
    return None


def _is_date_type(t: str) -> bool:
    t = (t or "").lower()
    return "date" in t or "time" in t or t == "timestamp"


def _is_numeric_type(t: str) -> bool:
    t = (t or "").lower()
    return "int" in t or "decimal" in t or "float" in t or "double" in t


def get_process_data_table(conn) -> str:
    import os
    fixed = os.getenv("PROCESS_TABLE_NAME", "").strip()
    return fixed or "preprocessing"


def _find_date_column(columns: list[dict]) -> str | None:
    date_names = ["timestamp", "date", "created_at", "recorded_at", "dt", "time", "날짜"]
    found = _pick_column(columns, date_names)
    if found:
        return found
    for c in columns:
        if _is_date_type(c["type"]):
            return c["name"]
    return None


def get_process_column_map(conn, table_name: str) -> dict:
    columns = get_columns(conn, table_name)
    date_col = _find_date_column(columns)
    quantity_col = _pick_column(columns, ["quantity", "amount", "count", "qty", "output", "생산", "수량"])
    pass_rate_col = _pick_column(columns, ["pass_rate", "pass", "quality", "ok_rate", "양품률", "품질"])
    defect_col = _pick_column(columns, ["quality_defect", "defect", "defect_rate", "fail", "ng", "불량"])
    consumption_col = _pick_column(columns, ["consumption", "usage", "kwh", "energy", "power", "에너지", "소비"])
    efficiency_col = _pick_column(columns, ["efficiency", "uptime", "oee", "rate", "효율", "가동률"])
    line_col = _pick_column(columns, ["line", "line_id", "line_name", "라인", "공정"])
    lot_col = _pick_column(columns, ["lot_id", "lot", "batch", "lot_no", "batch_id", "LOT", "id"])
    result_col = _pick_column(
        columns,
        [
            "quality_defect", "y_defect", "result", "pass_fail", "judge", "judgment",
            "판정", "합불", "ok_ng", "pass_fail_result", "quality_result", "judgement",
        ],
    )
    numeric_cols = [c["name"] for c in columns if _is_numeric_type(c["type"])]
    quantity_fallback = quantity_col or next((n for n in numeric_cols if "pass" not in n.lower() and "rate" not in n.lower() and "quality" not in n.lower()), None) or (numeric_cols[0] if numeric_cols else None)
    return {
        "table": table_name,
        "dateCol": date_col,
        "quantityCol": quantity_fallback,
        "passRateCol": pass_rate_col,
        "defectCol": defect_col,
        "consumptionCol": consumption_col,
        "efficiencyCol": efficiency_col,
        "lineCol": line_col,
        "lotCol": lot_col,
        "resultCol": result_col,
        "numericCols": numeric_cols,
    }


def get_today_date_string() -> str:
    try:
        tz = ZoneInfo(BACKEND_DATE_TZ)
        now = datetime.now(tz)
        return now.strftime("%Y-%m-%d")
    except Exception:
        return datetime.utcnow().strftime("%Y-%m-%d")


def get_dashboard_date_strings() -> dict:
    today_str = get_today_date_string()
    y, m, d = [int(x) for x in today_str.split("-")]
    from calendar import monthrange
    from datetime import date, timedelta
    last_day = monthrange(y, m)[1]
    first_of_month = f"{y}-{m:02d}-01"
    last_of_month_str = f"{y}-{m:02d}-{last_day:02d}"
    today = date(y, m, d)
    # 월~일: Python weekday() = 월0 .. 일6
    mon = today - timedelta(days=today.weekday())
    sun = mon + timedelta(days=6)
    week_start_str = mon.strftime("%Y-%m-%d")
    week_end_str = sun.strftime("%Y-%m-%d")
    return {
        "todayStr": today_str,
        "weekStartStr": week_start_str,
        "weekEndStr": week_end_str,
        "firstOfMonth": first_of_month,
        "lastOfMonthStr": last_of_month_str,
    }
